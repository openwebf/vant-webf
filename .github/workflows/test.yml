name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
