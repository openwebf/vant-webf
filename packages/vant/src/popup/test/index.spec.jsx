import { nextTick } from 'vue';
import { mount, triggerDrag } from '../../../test';
import { Popup } from '..';

let wrapper;
afterEach(() => {
  wrapper.unmount();
});

test('should lazy render content by default', async () => {
  wrapper = mount(Popup, {
    slots: {
      default: () => <div class="foo" />,
    },
  });

  expect(wrapper.find('.foo').exists()).toBeFalsy();

  await wrapper.setProps({ show: true });
  expect(wrapper.find('.foo').exists()).toBeTruthy();
});

test('should change z-index when using z-index prop', async () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      zIndex: 10,
    },
  });

  await nextTick();
  expect(wrapper.find('.van-popup').style.zIndex).toEqual('10');
  expect(wrapper.find('.van-overlay').style.zIndex).toEqual('10');
});

test('should lock scroll when showed', async () => {
  wrapper = mount(Popup);
  expect(document.body.classList.contains('van-overflow-hidden')).toBeFalsy();

  await wrapper.setProps({ show: true });
  expect(document.body.classList.contains('van-overflow-hidden')).toBeTruthy();
});

test('should lock page scroll by default', () => {
  const wrapper1 = mount(Popup, {
    props: {
      show: true,
    },
  });
  expect(document.body.classList.contains('van-overflow-hidden')).toBeTruthy();
  triggerDrag(document, 0, 100);
  triggerDrag(document, 0, -150);

  const wrapper2 = mount(Popup, {
    props: {
      show: true,
    },
  });
  wrapper1.unmount();
  expect(document.body.classList.contains('van-overflow-hidden')).toBeTruthy();

  wrapper2.unmount();
  expect(document.body.classList.contains('van-overflow-hidden')).toBeFalsy();
});

test('should allow to using teleport prop', async () => {
  const div = document.createElement('div');
  mount({
    render() {
      return <Popup show teleport={div} />;
    },
  });

  expect(div.querySelector('.van-popup')).toBeTruthy();
});

test('should render overlay by default', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
    },
  });
  expect(wrapper.find('.van-overlay').exists()).toBeTruthy();
});

test('should not render overlay when overlay prop is false', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      overlay: false,
    },
  });
  expect(wrapper.find('.van-overlay').exists()).toBeFalsy();
});

test('should emit click-overlay event when overlay is clicked', async () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
    },
  });
  const overlay = wrapper.find('.van-overlay');
  overlay.trigger('click');
  expect(wrapper.emitted('click-overlay')).toHaveLength(1);
});

test('should emit open event when show prop is set to true', async () => {
  const onOpen = jest.fn();
  wrapper = mount(Popup, {
    props: {
      onOpen,
    },
  });

  await wrapper.setProps({ show: true });
  expect(onOpen).toHaveBeenCalledTimes(1);
});

test('should emit close event when show prop is set to false', async () => {
  const onClose = jest.fn();
  wrapper = mount(Popup, {
    props: {
      show: true,
      onClose,
    },
  });

  await wrapper.setProps({ show: false });
  expect(onClose).toHaveBeenCalledTimes(1);
});

test('should emit close event after clicking the overlay', async () => {
  const onClose = jest.fn();
  wrapper = mount(Popup, {
    props: {
      show: true,
      onClose,
      'onUpdate:show': (show) => {
        wrapper.setProps({ show });
      },
    },
  });

  await wrapper.find('.van-overlay').trigger('click');
  expect(onClose).toHaveBeenCalledTimes(1);
});

test('should change duration when using duration prop', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      duration: 0.5,
    },
  });

  const popup = wrapper.find('.van-popup').element;
  const overlay = wrapper.find('.van-overlay').element;
  expect(popup.style.animationDuration).toEqual('0.5s');
  expect(overlay.style.animationDuration).toEqual('0.5s');
});

test('should have "van-popup--round" class when setting the round prop', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      round: true,
    },
  });

  expect(wrapper.find('.van-popup--round').exists()).toBeTruthy();
});

test('should render close icon when using closeable prop', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      closeable: true,
    },
  });

  wrapper.find('.van-popup__close-icon').trigger('click');
  expect(wrapper.emitted('update:show')[0][0]).toEqual(false);
});

test('should emit click-close-icon event when close icon is clicked', () => {
  wrapper = mount(Popup, {
    propsData: {
      show: true,
      closeable: true,
    },
  });

  wrapper.find('.van-popup__close-icon').trigger('click');
  expect(wrapper.emitted('click-close-icon')).toHaveLength(1);
});

test('should render correct close icon when using close-icon prop', () => {
  wrapper = mount(Popup, {
    props: {
      show: true,
      closeable: true,
      closeIcon: 'success',
    },
  });

  expect(wrapper.find('.van-popup__close-icon').html()).toMatchSnapshot();
});

test('should change icon class prefix when using icon-prefix prop', () => {
  const wrapper = mount(Popup, {
    props: {
      show: true,
      closeable: true,
      iconPrefix: 'my-icon',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render overlay-content slot correctly', () => {
  const wrapper = mount(Popup, {
    props: {
      show: true,
    },
    slots: {
      'overlay-content': () => 'Custom Overlay Content',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should allow to prevent close with before-close prop', async () => {
  const wrapper = mount(Popup, {
    props: {
      show: true,
      beforeClose: () => false,
    },
  });

  await wrapper.find('.van-overlay').trigger('click');
  expect(wrapper.emitted('update:show')).toBeFalsy();

  await wrapper.setProps({ beforeClose: () => true });
  await wrapper.find('.van-overlay').trigger('click');
  expect(wrapper.emitted('update:show')[0]).toEqual([false]);
});

test('should not call before-close when show prop becomes false', async () => {
  const beforeClose = jest.fn();
  const wrapper = mount(Popup, {
    props: {
      show: true,
      beforeClose,
    },
  });

  await wrapper.setProps({ show: false });
  expect(beforeClose).toHaveBeenCalledTimes(0);
});

test('should have safe-area-inset-top class when using safe-area-inset-top prop', () => {
  const wrapper = mount(Popup, {
    props: {
      show: true,
      safeAreaInsetTop: true,
    },
  });

  expect(wrapper.find('.van-popup').classes()).toContain('van-safe-area-top');
});

test('should have safe-area-inset-bottom class when using safe-area-inset-bottom prop', () => {
  const wrapper = mount(Popup, {
    props: {
      show: true,
      safeAreaInsetBottom: true,
    },
  });

  expect(wrapper.find('.van-popup').classes()).toContain(
    'van-safe-area-bottom'
  );
});
