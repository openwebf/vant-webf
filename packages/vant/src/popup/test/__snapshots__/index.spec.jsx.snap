// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change icon class prefix when using icon-prefix prop 1`] = `
<transition-stub role="button"
                 tabindex="0"
>
  <div class="van-overlay">
  </div>
</transition-stub>
<transition-stub>
  <div role="dialog"
       tabindex="0"
       class="van-popup van-popup--center"
  >
    <i class="van-badge__wrapper my-icon my-icon-cross van-popup__close-icon van-popup__close-icon--top-right van-haptics-feedback"
       role="button"
       tabindex="0"
    >
    </i>
  </div>
</transition-stub>
`;

exports[`should render correct close icon when using close-icon prop 1`] = `
<i class="van-badge__wrapper van-icon van-icon-success van-popup__close-icon van-popup__close-icon--top-right van-haptics-feedback"
   role="button"
   tabindex="0"
>
</i>
`;

exports[`should render overlay-content slot correctly 1`] = `
<transition-stub role="button"
                 tabindex="0"
>
  <div class="van-overlay">
    Custom Overlay Content
  </div>
</transition-stub>
<transition-stub>
  <div role="dialog"
       tabindex="0"
       class="van-popup van-popup--center"
  >
  </div>
</transition-stub>
`;
